compiler/src/dotty/tools/dotc/transform/CheckShadowing.scala

```diff
@@ -0,0 +1,312 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.Trees.EmptyTree
+import dotty.tools.dotc.transform.MegaPhase
+import dotty.tools.dotc.transform.MegaPhase.MiniPhase
+import dotty.tools.dotc.report
+import dotty.tools.dotc.core.Contexts.*
+import dotty.tools.dotc.core.Flags.*
+import dotty.tools.dotc.util.{Property, SrcPos}
+import dotty.tools.dotc.core.Symbols.ClassSymbol
+import dotty.tools.dotc.core.Names.Name
+import dotty.tools.dotc.core.Symbols.Symbol
+import dotty.tools.dotc.core.Flags.EmptyFlags
+import dotty.tools.dotc.ast.tpd.TreeTraverser
+import dotty.tools.dotc.core.Types.watchList
+import dotty.tools.dotc.core.Types.NoType
+import dotty.tools.dotc.core.Types.Type
+import dotty.tools.dotc.core.Types
+import dotty.tools.dotc.semanticdb.TypeOps
+import dotty.tools.dotc.cc.boxedCaptureSet
+import dotty.tools.dotc.core.Symbols.NoSymbol
+import dotty.tools.dotc.transform.SymUtils.isParamOrAccessor
+import scala.collection.mutable
+import dotty.tools.dotc.core.Scopes.Scope
+import scala.collection.immutable.HashMap
+import dotty.tools.dotc.core.Symbols
+import dotty.tools.dotc.typer.ImportInfo
+import dotty.tools.dotc.ast.untpd.ImportSelector
+import dotty.tools.dotc.core.StdNames.nme
+import dotty.tools.dotc.ast.untpd
+import dotty.tools.dotc.core.Denotations.SingleDenotation
+import dotty.tools.dotc.ast.Trees.Ident
+import dotty.tools.dotc.core.Names.TypeName
+import dotty.tools.dotc.core.Names.TermName
+import dotty.tools.dotc.core.Mode.Type
+import dotty.tools.dotc.core.Names.SimpleName
+
+class CheckShadowing extends MiniPhase:
```

Comment:
There is quite a lot of logic. The same is true for `CheckUnused`. We should've already run the `pos` tests that pass linting with all the flags enabled. I think it's a requirement for moving forward, as this code could even crash for some cases we didn't consider (and that's what happened with `CheckUnused`, so we had to spend a lot of time fixing it). There is a file `CompilationTests` (`compiler/test/dotty/tools/dotc/CompilationTests.scala`) where all the compilation tests are specified. We should take all the tests from `pos` and run them with linting while ignoring warnings (some will be generated, but they can be kept as non-fatal). We want to search for crashes.
```scala
compileFilesInDir("tests/pos", defaultOptions.and("-Xlint:private-shadow", "-Xlint:type-parameter-shadow", "-Wunused"))
```

Reply:
Thank you for that precious proposal, I have made the necessary changes in `CompilationTests` to detect compiler crashes
 and had to resolve one issue with the synthetic symbol generated for the case class "copy" method.

On a second hand, I turned on the fatal-warnings flag as suggested and checked for every false-positives. I encountered only one case which was due to an `export` clause : `TypeDef` `Trees` were reached from the export site, and some types not accessible from there were taken wrongly into account.

There is the list of all tests triggering a warning for -Xlint:private-shadow or -Xlint:type-parameter-shadow, checked as true positives :
[listShadowingTruePositives.txt](https://github.com/lampepfl/dotty/files/12727831/listShadowingTruePositives.txt)